src/odometry/LeggedOdometryManager.cpp

```diff
@@ -0,0 +1,705 @@
+
+namespace so = stateObservation;
+
+namespace mc_state_observation::odometry
+{
+
+///////////////////////////////////////////////////////////////////////
+/// -------------------------Legged Odometry---------------------------
+///////////////////////////////////////////////////////////////////////
+
+void LeggedOdometryManager::init(const mc_control::MCController & ctl,
+                                 Configuration odomConfig,
+                                 ContactsManagerConfiguration contactsConf)
+
+{
+  robotName_ = odomConfig.robotName_;
+  odometryType_ = odomConfig.odometryType_;
+  withYawEstimation_ = odomConfig.withYaw_;
+  velocityUpdate_ = odomConfig.velocityUpdate_;
+  odometryName_ = odomConfig.odometryName_;
+  const auto & robot = ctl.robot(robotName_);
+  odometryRobot_ = mc_rbdyn::Robots::make();
+  odometryRobot_->robotCopy(robot, "odometryRobot");
+
+  fbPose_.translation() = robot.posW().translation();
+  fbPose_.rotation() = robot.posW().rotation();
+
+  contactsManager_.init(ctl, robotName_, contactsConf);
+
+  if(!ctl.datastore().has("KinematicAnchorFrame::" + ctl.robot(robotName_).name()))
+  {
+    double leftFootRatio = robot.indirectSurfaceForceSensor("LeftFootCenter").force().z()
+                           / (robot.indirectSurfaceForceSensor("LeftFootCenter").force().z()
+                              + robot.indirectSurfaceForceSensor("RightFootCenter").force().z());
+
+    worldAnchorPose_ = conversions::kinematics::fromSva(
+        sva::interpolate(robot.surfacePose("RightFootCenter"), robot.surfacePose("LeftFootCenter"), leftFootRatio),
+        so::kine::Kinematics::Flags::pose);
+  }
+  else
+  {
+    worldAnchorPose_ = conversions::kinematics::fromSva(
+        ctl.datastore().call<sva::PTransformd>("KinematicAnchorFrame::" + ctl.robot(robotName_).name(),
+                                               ctl.robot(robotName_)),
+        so::kine::Kinematics::Flags::pose);
+  }
+
+  auto & logger = (const_cast<mc_control::MCController &>(ctl)).logger();
+  logger.addLogEntry(odometryName_ + "_odometryRobot_posW",
+                     [this]() -> sva::PTransformd { return odometryRobot().posW(); });
+
+  logger.addLogEntry(odometryName_ + "_odometryRobot_velW",
+                     [this]() -> sva::MotionVecd { return odometryRobot().velW(); });
+
+  logger.addLogEntry(odometryName_ + "_odometryRobot_accW",
+                     [this]() -> sva::MotionVecd { return odometryRobot().accW(); });
+  if(odomConfig.withModeSwitchInGui_)
+  {
+    ctl.gui()->addElement({odometryName_, "Odometry"},
+                          mc_rtc::gui::ComboInput(
+                              "Choose from list", {"6D", "Flat"},
+                              [this]() -> std::string
+                              {
+                                if(odometryType_ == measurements::OdometryType::Flat) { return "Flat"; }
+                                else { return "6D"; }
+                              },
```

Comment:
You need a function for this, it's already implemented twice in this PR and I can count 7 or 8 more in the ObserversPR branch :D